File: src/circular_progress.cpp
#include "circular_progress.h"

CircularProgress::CircularProgress(QWidget *parent) : QWidget(parent)
{
	this->resize(this->width,this->height);
}

void CircularProgress::paintEvent(QPaintEvent *)
{
	// set progress values
	int width 	= this->width - this->progWidth;
	int height 	= this->height - this->progWidth;
	int margin 	= this->progWidth / 2;
	int value 	= this->value * 360 / this->maxValue;

	// create painter
	QPainter painter(this);
	painter.setRenderHint(QPainter::Antialiasing);

	// create rectangle
	QRectF rect = QRectF(0,0,this->width,this->height);
	painter.setPen(Qt::NoPen);
	painter.drawRect(rect);
	painter.setFont(QFont("Segoe UI", this->fontSize));

	// create pen
	QPen pen = QPen();
	pen.setColor(this->progColor);
	pen.setWidth(this->progWidth);
	// round cap
	if (this->progRoundedCap)
		pen.setCapStyle(Qt::RoundCap);
	
	// create arc
	painter.setPen(pen);
	painter.drawArc(margin, margin, width, height, -90*16, -value*16); // 360 degreed = 5760 in Qt

	// create text
	QString percent = QString::fromStdString(std::to_string(this->value) + this->suffix);
	painter.drawText(rect, Qt::AlignCenter, percent);

	// close painter
	painter.end();
}

void CircularProgress::setValue(int newValue)
{
	// change progress value
	this->value = newValue;
	// repaint bar
	this->update();
}

void CircularProgress::addShadow(bool enable)
{
	if (enable)
	{
		this->shadow->setBlurRadius(15);
		this->shadow->setXOffset(0);
		this->shadow->setYOffset(0);
		this->shadow->setColor(QColor(0,0,0,120));
		this->setGraphicsEffect(this->shadow);
	}
}

File: ui/circular_progress/circular_progress.py
from PySide2.QtCore import *
from PySide2.QtGui import *
from PySide2.QtWidgets import *

class CircularProgress(QWidget):
    def __init__(self):
        QWidget.__init__(self)
        # progress properties
        self.value              = 0
        self.width              = 200
        self.height             =200
        self.progressWidth      = 10
        self.progressRoundedCap = True
        self.progressColor      = 0x4988D1
        self.maxValue           = 100
        self.fontFamily         = "Segoe UI"
        self.fontSize           = 12
        self.suffix             = "%"
        self.textColor          = 0x4988D1
        self.enableShadow       = True

        # set default size
        self.resize(self.width, self.height)

    # set progress value
    def setValue(self, value):
        self.value = value
        self.repaint()

    # draw shadow
    def addShadow(self, enable):
        if enable:
            self.shadow = QGraphicsDropShadowEffect(self)
            self.shadow.setBlurRadius(15)
            self.shadow.setXOffset(0)
            self.shadow.setYOffset(0)
            self.shadow.setColor(QColor(0,0,0,120))
            self.setGraphicsEffect(self.shadow)

    # paint event definition, used by qt to render widget
    def paintEvent(self, event):
        # set progress values
        width = self.width - self.progressWidth
        height = self.height - self.progressWidth
        margin = self.progressWidth / 2
        value  = self.value * 360 / self.maxValue

        # make painter
        paint = QPainter()
        paint.begin(self)
        paint.setRenderHint(QPainter.Antialiasing)
        paint.setFont(QFont(self.fontFamily, self.fontSize))

        # create rectangle
        rect = QRect(0, 0, self.width, self.height)
        paint.setPen(Qt.NoPen)
        paint.drawRect(rect)

        # create pen
        pen = QPen()
        pen.setColor(QColor(self.progressColor))
        pen.setWidth(self.progressWidth)
        if self.progressRoundedCap:
            pen.setCapStyle(Qt.RoundCap)

        # draw arc
        paint.setPen(pen)
        paint.drawArc(margin, margin, width, height, -90 * 16, -value * 16)

        # create text
        pen.setColor(QColor(self.textColor))
        paint.setPen(pen)
        paint.drawText(rect, Qt.AlignCenter, f"{self.value}{self.suffix}")

        # close painter
        paint.end()

